Convert project to kubebuilder version 2

chuckha opened the pull request:
Signed-off-by: Chuck Ha <chuckh@vmware.com>

**What this PR does / why we need it**:
This PR converts capa to a kubebuilder v2 project.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #989 

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

vincepri commented on cmd/manager/main.go:
```diff
@@ -19,78 +19,68 @@ package main
 
+var (
+	myscheme = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
```
Are these coming from kubebuilder?

chuckha replied:
it's `scheme` by default but that shadows the import so I changed it to `myscheme` instead.

The setupLog is there by default in v2 main.go though.

vincepri replied:
We should probably open an issue in kubebuilder

vincepri commented on cmd/manager/main.go:
```diff
@@ -19,78 +19,68 @@ package main
 
-	// Setup controller-runtime logger.
-	log.SetLogger(klogr.New())
+	ctrl.SetLogger(zap.Logger(true))
```
I thought we were sticking to klogr for the time being, has that changed?

chuckha replied:
yeah, this one snuck through, reverting

vincepri commented on hack/tools/tools.go:
```diff
@@ -0,0 +1,35 @@
+	_ "k8s.io/code-generator/cmd/informer-gen"
+	_ "k8s.io/code-generator/cmd/lister-gen"
+	_ "k8s.io/code-generator/cmd/register-gen"
+	_ "k8s.io/code-generator/cmd/set-gen"
```
We can probably remove code-generator from here and use `controller-gen` to generate the deepcopy files

chuckha replied:
This shouldn't actually be here, I'm not making this change as it's too big. I'll remove this entirely

vincepri commented on pkg/apis/infrastructure/v1alpha2/awscluster_types.go:
```diff
@@ -42,77 +38,6 @@ type AWSClusterSpec struct {
 	SSHKeyName string `json:"sshKeyName,omitempty"`
 }
 
-// NetworkSpec encapsulates all things related to AWS network.
```
Why move these? Seems they are only used within this file

chuckha replied:
these got moved to keep the awscluster_types.go and awsmachine_types.go files to contain only the kubebuilder generated types. The supporting types live in types.go

vincepri commented on pkg/apis/infrastructure/v1alpha2/zz_generated.deepcopy.go:
```diff
@@ -16,20 +16,19 @@ See the License for the specific language governing permissions and
 */
 
-// Code generated by main. DO NOT EDIT.
+// autogenerated by controller-gen object, do not modify manually; DO NOT EDIT
```
How were these generated with controller-gen? I don't see any changes in the Makefile / bazel

chuckha replied:
controller-gen uses the PROJECT file to detect which version of kubebuilder to use. I updated that to version 2 so no changes to the build need to happen.

vincepri replied:
That makes sense, but how was this file generated? Using which command?

chuckha replied:
oh, I see what you're saying, yes I'll update the makefile, it's using the normal kubebuilder v2 controller-gen command with a modified path

vincepri commented on cmd/manager/main.go:
```diff
@@ -19,78 +19,68 @@ package main
 import (
```
Should this file be moved in the root directory?

chuckha replied:
no, I don't think so. Since we are shipping multiple binaries with this project I think it makes more sense organizationally to keep them all under `cmd/<binary name>/`

vincepri replied:
Wouldn't kubebuilder complain when generating new controllers or types if it can't find a main.go in the root?

chuckha replied:
Yep it does, but I'm not terribly concerned by it. Generating new apis put them in `api/` instead of `/pkg/apis/`.

Moving main.go is going to require changing paths and some Bazel. I can see this task falling into the scope of this PR. I'd prefer doing it after due to the build changes, but if you'd like it done in this PR that's fine too.

vincepri replied:
That's fine, I was mostly wondering if it was on your radar, rather than pushing it for doing it in the same PR

detiber commented on cmd/manager/main.go:
```diff
@@ -19,78 +19,68 @@ package main
 	}
 
-	// Setup a Manager
-	syncPeriod := 10 * time.Minute
```
This likely needs to be preserved to avoid the default going back to 10 hours.

chuckha replied:
ok yeah that's fair

chuckha commented:
@vincepri @detiber this is ready for another pass

chuckha commented on cmd/manager/main.go:
```diff
@@ -19,78 +19,71 @@ package main
 import (
 	"flag"
 	"net/http"
-	_ "net/http/pprof"
```
this should not go away

chuckha commented on cmd/manager/main.go:
```diff
@@ -19,78 +19,71 @@ package main
+		MetricsBindAddress: metricsAddr,
+		LeaderElection:     enableLeaderElection,
 		SyncPeriod:         &syncPeriod,
-		Namespace:          *watchNamespace,
```
should not go away

chuckha commented on cmd/manager/main.go:
```diff
@@ -19,78 +19,71 @@ package main
-	}
-
-	// Initialize event recorder.
-	record.InitFromRecorder(mgr.GetEventRecorderFor("aws-controller"))
```
also still needed

detiber replied:
Should it, or is it something that we should set in the individual controllers instead?

chuckha replied:
oh i'm not sure. will it still work initialized in main? I'd like to keep it there and then move it later

detiber commented on pkg/apis/infrastructure/v1alpha2/awsmachinetemplate_types.go:
```diff
@@ -1,63 +0,0 @@
-/*
```
Doh, I think the AWSMachineTemplate type should probably have been preserved.

chuckha replied:
Oh yeah. Totally.

detiber commented on pkg/cloud/scope/cluster.go:
```diff
@@ -127,13 +127,6 @@ func (s *ClusterScope) ControlPlaneConfigMapName() string {
-		clusterv1alpha2.MachineClusterLabelName: s.Cluster.Name,
-	})
-}
-
```
I'm guessing this wasn't used anywhere?

chuckha replied:
That's correct, wasn't used

vincepri commented on cmd/manager/main.go:
```diff
@@ -19,78 +19,77 @@ package main
+}
+
 func main() {
-	klog.InitFlags(nil)
```
Don't we need this?

chuckha replied:
yes. definitely.

chuckha commented on config/crd/bases/infrastructure.cluster.x-k8s.io_awsclusters.yaml:
```diff
@@ -11,8 +11,6 @@ spec:
     plural: awsclusters
   scope: ""
-  subresources:
-    status: {}
```
sure is, let me regen. I probably missed some of those kubebuilder markers i didn't know needed to exist. Let me add those.

chuckha commented:
@detiber @vincepri 

This gets us to the point of the un-implemented controller.

Take a look at config/ as well. I added some stuff to make the credentials secret (also look at the Makefile).

Otherwise the biggest changes: i deleted the examples directory in its entirety for v1alpha2.

detiber commented on config/manager/manager.yaml:
```diff
@@ -2,85 +2,30 @@ apiVersion: v1
   name: controller-manager
   namespace: system
   labels:
     control-plane: controller-manager
```
This should likely use a different value, otherwise it will conflict with other kubebuilder projects using the defaults in the same namespace.

chuckha replied:
Agreed.
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1016

detiber commented on Makefile:
```diff
@@ -87,24 +96,14 @@ e2e-janitor:
 	docker build --pull . -t $(MANAGER_IMAGE)
+	# TODO: sed probably needs to be gnu sed not bsd sed
+	sed -i '' -e 's@image: .*@image: '"${MANAGER_IMAGE}"'@' ./config/default/manager_image_patch.yaml
+	sed -i '' -e 's@credentials: .*@credentials: '"${B64_CREDENTIALS}"'@' ./config/default/aws_credentials_patch.yaml
```
Should this be part of the `docker-build` target? I would expect it as part of `generate-manifests` instead.

chuckha replied:
This is how kubebuilder v2 makefiles are.

I believe the idea behind this being part of the docker-build step is that you'd build for a release/deployment. At which point the sed is applied to the image patch and you can now deploy your app.

If it were part of the generate-manifest we would only be guessing at the name of the image. So it might look correct but it wouldn't be valid until the docker-build/push step anyway

detiber replied:
can't say I'm a fan, but we can go with it for now and reassess it later.

chuckha replied:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1014

detiber commented on BUILD.bazel:
```diff
@@ -80,3 +81,74 @@ filegroup(
+    tags = ["manual"],
+)
+
+stateful_set_patch(
```
This should likely be renamed since we are no longer using a StatefulSet

chuckha replied:
Agreed. I think figuring out what to do with Bazel is next on the list of cleanups here.

https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1015

detiber commented on BUILD.bazel:
```diff
@@ -80,3 +81,74 @@ filegroup(
+    "$(MANAGER_IMAGE_TAG)",
+]
+
+images = {
```
Should we have both a Dockerfile and a Bazel build for images? We should probably consolidate on one or the other.

chuckha replied:
I agree but every time I touch bazel my PR blows up. Do you think we can do this after this PR? I would be more than happy to make an issue for consolidating the build/release practice.

detiber replied:
Sounds good, lets handle this later

chuckha replied:
Added a comment https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1015#issuecomment-521278295

detiber commented:
lgtm, but would like @vincepri to take a look since he's doing similar work in cluster-api

vincepri commented:
Taking a look now
